two phases flow codes/Dambreak_LevelSet_script.py

```diff
@@ -0,0 +1,239 @@
+"""
+Dam Break by yiming liu
+
+2021/5/28
+"""
+
+from dolfin import *
+from dolfin.cpp.mesh import cells
+from mshr import *
+from ufl import indices
+import matplotlib.pyplot as plt
+#%matplotlib inline
+import numpy as np
+#from math import exp
+import time
+
+class CharacterLevelSetFunction(UserExpression):
+	'''
+	initialize the level set function
+	Firstly, determine the position of an arbitrary point in the domain
+	Secondly, calculate the distance to the minimum distance to the interface
+	Finally, calculate the CharacterLevelSetFunction, 
+	in the interface =1, out the interface = 0, on the transition area of interface it is continuous
+	'''
+	def __init__(self, param1, param2, param3, **kwargs):
+		super().__init__(**kwargs)
+		self.param1 = param1 # x boundary of interface
+		self.param2 = param2 # y boundary of interface
+		self.param3 = param3 # epsilon: thickness parameter
+
+	def eval(self, values, x):
+		if x[0] <= self.param1 and x[1] <= self.param2:
+			phi_d = - min(self.param1 - x[0], self.param2 - x[1])
+			values[0] = 1/(1+exp(phi_d/self.param3))
+		if x[0] > self.param1 and x[1] > self.param2:
+			dx=x[0]-self.param1
+			dy=x[1]-self.param2
+			phi_d = pow(pow(dx,2)+pow(dy,2),0.5)
+			values[0] = 1/(1+exp(phi_d/self.param3))
+		if x[0] > self.param1 and x[1] <= self.param2:
+			phi_d = x[0] - self.param1
+			values[0] = 1/(1+exp(phi_d/self.param3))
+		if x[0] <= self.param1 and x[1] > self.param2:
+			phi_d = x[1] - self.param2
+			values[0] = 1/(1+exp(phi_d/self.param3))
+
+processID =  MPI.comm_world.Get_rank()
```

Comment:
You are already running the script in parallel? Is this necessary and did you check that the script actually runs faster?

Reply:
Yes, I have already run this in parallel in the cluster 'benz'. When I choose 4-8 processors, I feel it is faster, but I did not check that, I can do this check later. 
```python
processID =  MPI.comm_world.Get_rank()

...

if processID == 0: print('...')  
```
This is mainly for printing only one set of parameters in the loop. If it is not recognized, the same line will be print multiple times in the terminal (the same as the number of selected processors)

Reply:
Okay I didn't know that you have an account on this machine. Did you get an introduction to it?

If you run your scripts using `mpirun`, you should be familiar with the MPI programming model. Your script is actually executed simultaneously using multiple processes. The processes will communicate with each other. This communication is mostly controlled by FEniCS.

Reply:
I didn't get an introduction. I was only told not to take up too many processor threads for a long time computing, and clean up the result files regularly to free up space. It would be better if there was an introduction, I could read it.